py/dml/c_backend.py

```diff
@@ -1977,24 +1964,53 @@ def init_trait_vtables_for_node(node, param_values, dims, parent_indices):
                         if name in method_overrides else "NULL")
                 elif member_kind == 'parameter':
                     # param_overrides contains C expression strings
-                    args.append(param_overrides[name])
+                    (value, t, indexed) = param_overrides[name]
+                    var = f'_param_{name}'
+                    if indexed:
+                        args.append(f'(void *)((uintptr_t){var} + 1)')
+                        array_type = t
+                        for d in reversed(node.dimsizes):
+                            array_type = TArray(array_type,
+                                                mkIntegerLiteral(node.site, d))
+                        param_decl.append(f'''\
+    {TPtr(array_type).declaration(var)} = malloc(sizeof(*{var}));
+    ''')
+                        param_init.append(f'(*{var}){param_indices} = {value};')
+                    else:
+                        args.append('({static %s __attribute__((aligned(2))); %s = %s; &%s;})'
+                                    % (t.declaration(var), var, value, var))
                 else:
                     assert member_kind == 'session'
                     session_node = node.get_component(name)
                     assert session_node.objtype in ('session', 'saved')
                     args.append('offsetof(%s, %s)' % (
                         crep.structtype(dml.globals.device),
-                        crep.cref_node(session_node, indices)))
+                        crep.cref_session(
+                            session_node,
+                            (mkIntegerConstant(node.site, 0, False),)
+                            * node.dimensions)))
             # initialize vtable instance
-            vtable_arg = '&%s%s' % (node.traits.vtable_cname(trait), index_str)
-            init_calls.append(
-                (dims, '_tinit_%s(%s);\n'
-                 % (trait.name, ', '.join([vtable_arg] + args))))
-    for subnode in node.get_components():
-        if isinstance(subnode, objects.CompositeObject):
-            init_calls.extend(init_trait_vtables_for_node(
-                subnode, param_values, dims, indices))
-    return init_calls
+            vtable_arg = f'&{node.traits.vtable_cname(trait)}'
+            init_call = ('_tinit_%s(%s);\n'
+                 % (trait.name, ', '.join([vtable_arg] + args)))
+
+            if param_init:
+                code = param_decl
+                close = []
+                for (i, d) in enumerate(node.dimsizes):
+                    indent = '    ' * i
```

Comment:
An alternative to doing indents manually is to pass some kind of signal to the caller so that it then calls `out` with `preindent` and `postindent` as appropriate.

Reply:
I think the reason I couldn't calling vanilla `out` directly was related to the loops over vtable arrays. With them gone, I think we can `out` directly.
(And regarding "some kind of signal": I think the proper way to do that would be to create the code as a string using a `output.StrOutput`)

Reply:
Done (and the reason it can be done is that we no longer need to split the function, thanks to O0)